Remove high contrast theme #3095

fedejeanne opened the pull request:
Fixes https://github.com/eclipse-platform/eclipse.platform.ui/issues/3095

## How to test
- Before checking out this PR, start the _Runtime Application_ and set the theme to _High Contrast_
- Checkout this PR
- Start the _Runtime Application_

**Expected outcome**
- The application should switch automatically to the default theme: _Light_.
- There should be no _High Contrast_ theme available in the preferences anymore

fedejeanne committed:
Remove high contrast theme #3095

Fixes https://github.com/eclipse-platform/eclipse.platform.ui/issues/3095

fedejeanne commented:
cc @vogella

eclipse-platform-bot committed:
Version bump(s) for 4.37 stream

iloveeclipse commented:
Wasn't there some user asking to keep it? I can't find the comment on original issue, but I believe there was one.
Could you ask on mailing lists before removing it?

fedejeanne commented:
I'll draft it until we decide how to proceed.

Open questions: do we disable theming when to OS is set to use _High Contrast_? ATM only Windows was capable of reporting if _High Contrast_ was enabled, but I know for a fact that Ubuntu has a _High Contrast_ option too and maybe even Mac does? If that is correct then we need to adapt the implementations of `Display::getHighContrast()` for Linux and Mac and decide what to do when it's activated.

vogella commented:
+1 for disabling theming if high contrast is enabled. This looks the best

fedejeanne committed:
Disable theming when high contrast mode is enabled #3095

Currently only working for Windows since Linux and Mac do not report
about high contrast mode being enabled/disabled.

Contributes to
https://github.com/eclipse-platform/eclipse.platform.ui/issues/3095

fedejeanne commented:
Commit e3f02aa35ec851e3ac1f1321b09641cb49a7a30a is an attempt to update the preferences dialog accordingly (disabling theming) when _high contrast_ is activated.

Considerations:
- Only works on Windows (since Linux and Mac do not report the _high contrast_ ATM)
- The UX is not ideal but IMO is pretty good considering that disabling theming is (even on `master`) a bit "buggy". If you for example enable/disable theming, apply and close but without restarting then the next time you open the _Appearance_ preference, the UI does **not** reflect the current status but the last saved status.

Feedback is welcome (the PR remains drafted though).

laeubi commented:
> Commit https://github.com/eclipse-platform/eclipse.platform.ui/commit/e3f02aa35ec851e3ac1f1321b09641cb49a7a30a is an attempt to update the preferences dialog accordingly (disabling theming) when high contrast is activated.

You might want to share a screenshot here to get more feedback.

akurtakov commented:
Adding some support for Linux based on (oversimplified) themeName.contains("HighContrast") should be fairly trivial and I can do it needed.

fedejeanne commented:
Here's a screenshot of the preferences when _High contrast_ is **enabled** (Windows)

<details><summary>High Contrast enabled (Windows)</summary>
<p>

<img width="853" height="364" alt="image" src="https://github.com/user-attachments/assets/581b31fb-6f22-4e9f-9c4d-cbcc0a8e211d" />

</p>
</details> 

The checkbox _Enable Theming_ is disabled (in color green) and unchecked.

fedejeanne commented:
> Adding some support for Linux based on (oversimplified) themeName.contains("HighContrast") should be fairly trivial and I can do it needed.

... but we want to remove the high contrast theme. How would `themeName.contains("HighContrast")` ever yield `true`, @akurtakov ?

akurtakov commented:
> > Adding some support for Linux based on (oversimplified) themeName.contains("HighContrast") should be fairly trivial and I can do it needed.
> 
> ... but we want to remove the high contrast theme. How would `themeName.contains("HighContrast")` ever yield `true`, @akurtakov ?

It will check Gtk theme name(https://wiki.archlinux.org/title/GTK)  not Eclipse theme name. If the Gtk theme name contains HighContrast SWT Display.isHighContrast will return true, otherwise it will return false.

fedejeanne commented:
> It will check Gtk theme name(https://wiki.archlinux.org/title/GTK) not Eclipse theme name.

🤯 

In that case, you could also check for "ContrastHigh" since there is one for Mate that's called "ContrastHighInverse" (and there might be more like that?)

akurtakov commented:
https://github.com/eclipse-platform/eclipse.platform.swt/pull/2372 - Gtk implementation

fedejeanne commented:
> [eclipse-platform/eclipse.platform.swt#2372](https://github.com/eclipse-platform/eclipse.platform.swt/pull/2372) - Gtk implementation

Thank you! One thing that just occurred to me is: does the mechanism to notify the user about a change in the "High Contrast" also work on Linux? On Windows, when I activate the _High Contrast_ setting (at OS level) I get this dialog asking me to restart:

<img width="500" height="136" alt="image" src="https://github.com/user-attachments/assets/3de6fa75-a822-4e67-9d0d-ae63f2756cee" />

Which comes from here:

https://github.com/eclipse-platform/eclipse.platform.ui/blob/4a2727c448db22f5ebf7fad65987b100007de72e/bundles/org.eclipse.ui.ide.application/src/org/eclipse/ui/internal/ide/application/IDEWorkbenchAdvisor.java#L318-L333

Does this also currently work on Linux, @akurtakov ? If not, it might be worth considering adding (in another PR).

akurtakov commented:
Detection of "high contrast" theme on startup works just fine on Linux. Runtime theme change doesn't seem to "fire" settings event on Gtk. That would have to be fixed on SWT side but it is not a blocker for this PR.

akurtakov commented:
> > [eclipse-platform/eclipse.platform.swt#2372](https://github.com/eclipse-platform/eclipse.platform.swt/pull/2372) - Gtk implementation
> 
> Thank you! One thing that just occurred to me is: does the mechanism to notify the user about a change in the "High Contrast" also work on Linux? On Windows, when I activate the _High Contrast_ setting (at OS level) I get this dialog asking me to restart:
> <img alt="image" width="500" height="136" src="https://private-user-images.githubusercontent.com/2205684/474381885-3de6fa75-a822-4e67-9d0d-ae63f2756cee.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ0ODQzMzIsIm5iZiI6MTc1NDQ4NDAzMiwicGF0aCI6Ii8yMjA1Njg0LzQ3NDM4MTg4NS0zZGU2ZmE3NS1hODIyLTRlNjctOWQwZC1hZTYzZjI3NTZjZWUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MDZUMTI0MDMyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDc4NWEyYTBiMDQ0OTBjNmU5ZjRjYTdiMTYzN2Y0YjEyY2I1ODI0MjVkZjkyNTM5NzY5ZmIyNzk3MTMxMjYzOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Rz1LepkzN98H98Yf5OnMXIZBapZRGYuk-PbO7pT4HrM">
> 
> Does this also currently work on Linux, @akurtakov ? If not, it might be worth considering adding (in another PR).

It works on Linux now with Latest I-build. IMO the PR is good to be merged.

fedejeanne commented:
> It works on Linux now with Latest I-build. IMO the PR is good to be merged.

I didn't have time to look into Mac yet and I got no answer from other Mac users. Maybe we could wait until M1?

akurtakov commented:
@Phillipus maybe you can provide some Mac feedback here?

Phillipus commented:
> @Phillipus maybe you can provide some Mac feedback here?

Isn't High Contrast a Windows only theme? I don't see it on Mac or Linux.

akurtakov commented:
<img width="3199" height="1999" alt="image" src="https://github.com/user-attachments/assets/9c4430d5-2aaf-4fd0-81c7-8037be150948" />
I have it on Linux.

Phillipus commented:
OK, it's not on Mac.

vogella commented:
Let's merge in this case

fedejeanne commented:
> Let's merge in this case

Ok.

FTR the check failure seems unrelated:

```
19:01:34.743 [ERROR] Failed to execute goal org.eclipse.tycho.extras:tycho-p2-extras-plugin:5.0.0-SNAPSHOT:compare-version-with-baselines (compare-attached-artifacts-with-release) on project org.eclipse.text: Baseline and reactor have the same fully qualified version, but different content
19:01:34.743 [ERROR] different
19:01:34.743 [ERROR]    META-INF/ECLIPSE_.RSA: present in baseline only
19:01:34.743 [ERROR]    META-INF/ECLIPSE_.SF: present in baseline only
19:01:34.743 [ERROR] -> [Help 1]
...
```

... but I see that the Code Analysis check started running again. I'll wait.

fedejeanne commented:
Same as [before](https://github.com/eclipse-platform/eclipse.platform.ui/pull/3138#issuecomment-3161303239), the check failure is unrelated.

Merging.

fedejeanne commented:
The [nightly tests](https://download.eclipse.org/eclipse/downloads/drops4/I20250806-1800/testResults.php) look good so I think it's safe to say: _Adiós, high contrast theme. One less outdated feature to maintain_ 👏 .

Thank you all for your contribution! 

IMO this is worthy of an N&N entry. Does anyone want to do the honors? :-)
